### PR TITLE
Bugfix/60 problem with kernel integral architecture

### PR DESCRIPTION
**Pull request by MLuchmann**
# Bugfix: Problem with kernel integral architecture

## Description

The current implementation of the Integral Kernel layer only manages to learn the target function when the target space is discretized by only one value (shape(y) == (batch-size, 1, y-dim)) as it is the case in examples/selfsupervised.ipynb. However, when the target space is discretized with a value larger than 1, the operator fails to learn the target function (at least for specific examples).

### Which issue does this PR tackle?

  - New implementation of Neural Network Kernel is more general and fixes the issue.

### How does it solve the problem?

  - The Neural Network Kernel of the previous implementation was constructed as:
  $k(x_i, y_j) = NN\left( (x_i - y_j)^2 \right)$.
  The issue described above was due to the symmetric properties of the example combined with the specific construction of the kernel. The input function in basics.ipnyb has the property $u(-x) = -u(x)$. One can show that this property combined with the construction of our kernel leads to:
  $v_{pred}(y_j) = \frac{1}{N} \sum_i k(x_i, y_j) u(x_i) = -v_{pred}(-y_j)$. Therefore, the operator could only properly learn mappings from asym. to asym. functions (or sym. to sym.). 
  - The new kernel implementation is more general and does not rely on a convolution ($x_i - y_i$).
  
### Note

I had to change one of the test cases because I changed the way the IntegralKernel method processes the inputs x and y. Previously extra dimensions were added via ```.squeeze()```. I deleted these lines. The idea is that the reshaping is the responsibillity of the given kernel and not the IntegralKernel operation.

### How are the changes tested?

  - Tested via tests/test_neuraloperator.py
  - Tested via tests/test_convolution.py

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [ ] Conventions: The branch follows the `feature/title-slug` convention.
- [ ] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [x] The PR solves the issue it claims to solve and only this one.
- [x] Changes are tested sufficiently and all tests pass.
- [x] Documentation is complete and well-written.
- [x] Changelog has been updated, if necessary.

### CODE REVIEW

File: src/continuity/operators/common.py [hunk 1]
```diff
@@ -87,47 +88,68 @@ def __init__(
     def forward(self, x):
         """Forward pass."""
         x = self.first_layer(x)
+        x = self.act(x)
         for layer in self.hidden_layers:
             x = layer(x)
         return self.last_layer(x)
 
 
 class NeuralNetworkKernel(torch.nn.Module):
```
Comment (samuelburbulla) [inline src/continuity/operators/common.py:97]: We should move the `NeuralNetworkKernel` to `integralkernel.py`.

File: src/continuity/operators/neuraloperator.py [hunk 2]
```diff
@@ -43,17 +43,16 @@ def __init__(
         super().__init__()
 
         self.shapes = shapes
-
         self.lifting = NaiveIntegralKernel(
-            NeuralNetworkKernel(kernel_width, kernel_depth),
+            NeuralNetworkKernel(shapes, kernel_width, kernel_depth),
```
Comment (samuelburbulla) [inline src/continuity/operators/neuraloperator.py:47]: I agree that we should pass the `shapes` to the kernel.

File: src/continuity/operators/common.py [hunk 3]
```diff
@@ -87,47 +88,68 @@ def __init__(
...
+        shapes: DatasetShapes,
         kernel_width: int,
         kernel_depth: int,
-    ):
+        act: Optional[torch.nn.Module] = None,
+    ) -> None:
         super().__init__()
+
+        self.shapes = shapes
+        self.input_shape = shapes.y.dim + shapes.x.num * shapes.x.dim
```
Comment (samuelburbulla) [inline src/continuity/operators/common.py:125]: The kernel should not depend on `x.num`.

File: src/continuity/operators/common.py [hunk 4]
```diff
@@ -87,47 +88,68 @@ def __init__(
...
         self.net = DeepResidualNetwork(
-            1,
-            1,
+            self.input_shape,
+            shapes.x.num,
```
Comment (samuelburbulla) [inline src/continuity/operators/common.py:128]: A kernel should output a matrix `R^{d_u \times d_v}`, independent of `x.num`.

File: src/continuity/operators/common.py [hunk 5]
```diff
@@ -87,47 +88,68 @@ def __init__(
...
-            Tensor of shape (...)
-        """
-        r = ((x - y) ** 2).sum(dim=-1)
-        output_shape = r.shape
-        r = r.reshape((-1, 1))
-        k = self.net(r)
-        k = k.reshape(output_shape)
-        return k
+        # shapes that can change for different forward passes
+        batch_size = x.shape[0]
```
Comment (samuelburbulla) [inline src/continuity/operators/common.py:136]: This batching should be part of the `IntegralKernel`, the kernel itself is independent of the number of xs or ys passed. See my abstract base class implementation for the target signature of a `Kernel`, and this `NeuralNetworkKernel` should follow this concept. Everything else moves to the `(Naive)IntegralKernel`. Maybe we should also add an abstract base class for an `IntegralKernel`?

File: src/continuity/operators/integralkernel.py [hunk 6]
```diff
@@ -5,8 +5,40 @@
 """
 
 import torch
+from abc import ABC
 from typing import Callable, Union
 from continuity.operators import Operator
+from continuity.data import DatasetShapes
+
+
+class Kernel(torch.nn.Module, ABC):
```
Comment (samuelburbulla) [inline src/continuity/operators/integralkernel.py:14]: This is the general concept of a kernel. NeuralNetworkKernel should implement this using a neural network.

**Review by samuelburbulla — CHANGES_REQUESTED** [review]
Looks good so far, but I think we have to outline the general concepts more clearly first.

What exactly is a `Kernel`? What is input/output of a kernel? 

Let's define an abstract base class first: I implemented the `Kernel` base class, I think this is what a kernel is. 
NeuralNetworkKernel should implement this signature (using an NN) and everything related to batching should be handled by the integral kernel. How does this sound to you?

We should do the same for the IntegralKernel, first defining a base class.

File: src/continuity/operators/integralkernel.py (hunk 6 above)
Comment (samuelburbulla) [reply src/continuity/operators/integralkernel.py:14]: Actually, maybe it makes sense to use batched inputs (num_x, x_dim), (num_y, y_dim) for more efficient implementation?

File: src/continuity/operators/common.py [hunk 7]
```diff
@@ -87,47 +88,68 @@ def __init__(
     def forward(self, x):
         """Forward pass."""
         x = self.first_layer(x)
+        x = self.act(x)
```
[the remark this answers is not in the document]
Comment (MLuchmann) [reply src/continuity/operators/common.py:90]: I added the suggested changes and opened another PR.

File: src/continuity/operators/common.py (hunk 1 above)
Comment (MLuchmann) [reply src/continuity/operators/common.py]: Good pont. Done.

File: src/continuity/operators/common.py [hunk 8]
```diff
@@ -87,47 +88,68 @@ def __init__(
...
         for layer in self.hidden_layers:
             x = layer(x)
         return self.last_layer(x)
 
 
 class NeuralNetworkKernel(torch.nn.Module):
-    """Neural network kernel.
+    """General Neural network kernel. The neural network is constructed as a simple
```
[the remark this answers is not in the document]
Comment (MLuchmann) [reply src/continuity/operators/common.py]: The doc string is updated now.

File: src/continuity/operators/common.py (hunk 3 above)
Comment (MLuchmann) [reply src/continuity/operators/common.py]: The kernel got reimplemented and does not depend on x.num any longer.

File: src/continuity/operators/integralkernel.py [hunk 9]
```diff
@@ -5,8 +5,128 @@
...
+        assert batch_size == y.shape[0]
+        x_num, y_num = x.shape[1], y.shape[1]
+
+        # shapes that are fixed
+        x_dim, y_dim = self.shapes.x.dim, self.shapes.y.dim
+        u_dim, v_dim = self.shapes.u.dim, self.shapes.v.dim
+
+        # In order to evaluate all kernel values k(x_i, y_j), we need all x and y combinations
+        # The following mimics the torch.meshgrid function (only works for 1d function)
+        # https://stackoverflow.com/questions/72782751/how-to-perform-torch-meshgrid-over-multiple-tensors-in-parallel
```
Comment (samuelburbulla) [inline src/continuity/operators/integralkernel.py:112]: What does it mean "only works for 1d function"? x.dim == y.dim == 1?

Comment (MLuchmann) [reply src/continuity/operators/integralkernel.py:112]: Sorry, the comment is misleading. I will delete the text in the brackets. I believe torch.meshgrid could have been used to create all combinations of $x_i$ and $y_j$, even though we do not necessary have a grid because $x$ and $y$ can be given at any position. However, we can't use it because the function only allows 1d tensors.

Comment (samuelburbulla) [reply src/continuity/operators/integralkernel.py]: I added a test for the NeuralNetworkKernel (because I was not sure if this was working for several dimensions) and it turned out I had to change the code a little bit. Now it works and is tested :)

**Review by samuelburbulla — APPROVED** [review]
